src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
```

Comment:
There are 3 per import, and then repeated endlessly for each object file. For:
```
import std.stdio;
void test() { }
```
the following are generated:
```
_D6object7__arrayZ      COMDAT flags=x0 attr=x10 align=x0
_D6object8__assertFiZv  COMDAT flags=x0 attr=x10 align=x0
_D6object15__unittest_failFiZv  COMDAT flags=x0 attr=x10 align=x0
_D4core4stdc6stdint7__arrayZ    COMDAT flags=x0 attr=x10 align=x0
_D4core4stdc6stdint8__assertFiZv        COMDAT flags=x0 attr=x10 align=x0
_D4core4stdc6stdint15__unittest_failFiZv        COMDAT flags=x0 attr=x10 align=x0
_D3std8typecons7__arrayZ        COMDAT flags=x0 attr=x10 align=x0
_D3std8typecons8__assertFiZv    COMDAT flags=x0 attr=x10 align=x0
_D3std8typecons15__unittest_failFiZv    COMDAT flags=x0 attr=x10 align=x0
_D3std6traits7__arrayZ  COMDAT flags=x0 attr=x10 align=x0
_D3std6traits8__assertFiZv      COMDAT flags=x0 attr=x10 align=x0
_D3std6traits15__unittest_failFiZv      COMDAT flags=x0 attr=x10 align=x0
_D3std4meta7__arrayZ    COMDAT flags=x0 attr=x10 align=x0
_D3std4meta8__assertFiZv        COMDAT flags=x0 attr=x10 align=x0
_D3std4meta15__unittest_failFiZv        COMDAT flags=x0 attr=x10 align=x0
```
and:
```
db      066h,06fh,06fh,000h,063h,03ah,05ch,063h ;foo.c:\c
db      062h,078h,05ch,06dh,061h,072h,073h,05ch ;bx\mars\
db      064h,072h,075h,06eh,074h,069h,06dh,065h ;druntime
db      05ch,069h,06dh,070h,06fh,072h,074h,05ch ;\import\
db      06fh,062h,06ah,065h,063h,074h,02eh,064h ;object.d
db      000h,063h,03ah,05ch,063h,062h,078h,05ch ;.c:\cbx\
db      06dh,061h,072h,073h,05ch,064h,072h,075h ;mars\dru
db      06eh,074h,069h,06dh,065h,05ch,069h,06dh ;ntime\im
db      070h,06fh,072h,074h,05ch,063h,06fh,072h ;port\cor
db      065h,05ch,073h,074h,064h,063h,05ch,073h ;e\stdc\s
db      074h,064h,069h,06eh,074h,02eh,064h,000h ;tdint.d.
db      063h,03ah,05ch,063h,062h,078h,05ch,06dh ;c:\cbx\m
db      061h,072h,073h,05ch,070h,068h,06fh,062h ;ars\phob
db      06fh,073h,05ch,073h,074h,064h,05ch,074h ;os\std\t
db      079h,070h,065h,063h,06fh,06eh,073h,02eh ;ypecons.
db      064h,000h,063h,03ah,05ch,063h,062h,078h ;d.c:\cbx
db      05ch,06dh,061h,072h,073h,05ch,070h,068h ;\mars\ph
db      06fh,062h,06fh,073h,05ch,073h,074h,064h ;obos\std
db      05ch,074h,072h,061h,069h,074h,073h,02eh ;\traits.
db      064h,000h,063h,03ah,05ch,063h,062h,078h ;d.c:\cbx
db      05ch,06dh,061h,072h,073h,05ch,070h,068h ;\mars\ph
db      06fh,062h,06fh,073h,05ch,073h,074h,064h ;obos\std
db      05ch,06dh,065h,074h,061h,02eh,064h,000h ;\meta.d.
```
used only by those helper functions.

The object file is 2,229 bytes, although the generated code for `test(){}` is one byte.